### PR TITLE
Update to Keycloak 25

### PR DESCRIPTION
**Pull request by xgp**
- Update Keycloak version to `25.0.0`
- Update java version to `21`, now the Keycloak standard per https://www.keycloak.org/2024/06/keycloak-2500-released.html#_java_21_support and https://www.keycloak.org/2024/06/keycloak-2500-released.html#_java_17_support_is_deprecated
- Change dependency from `keycloak-model-legacy` to `keycloak-model-storage` per https://github.com/keycloak/keycloak/issues/26657

### CODE REVIEW

**Comment by daniel-frak** [comment]
Hi! Thanks for the contribution!

For the changes to work, you will also need to update:

- [.env](https://github.com/daniel-frak/keycloak-user-migration/blob/master/docker/.env) (`KEYCLOAK_IMAGE`, `MAVEN_IMAGE` and `OPENJDK_IMAGE`)
- [maven.yml](https://github.com/daniel-frak/keycloak-user-migration/blob/master/.github/workflows/maven.yml) (the `Set up JDK 17` step)
- [release.yml](https://github.com/daniel-frak/keycloak-user-migration/blob/master/.github/workflows/release.yml) (the `Set Up Java` step)
- [README.md](https://github.com/daniel-frak/keycloak-user-migration/blob/master/README.md) (the `Compatibility history` table - the plugin version should be `SNAPSHOT`)

Here's a link to the "Keycloak version upgrade quickstart guide", for future reference:

https://github.com/daniel-frak/keycloak-user-migration/blob/master/CONTRIBUTING.md#quick-guide-to-upgrading-keycloak-compatibility

File: README.md [hunk 1]
```diff
@@ -22,7 +22,8 @@ https://codesoapbox.dev/keycloak-user-migration
 
 | Keycloak Version | Version/Commit                                                                                                                                     |
 |------------------|----------------------------------------------------------------------------------------------------------------------------------------------------|
-| 24.X             | [4.0.0](https://github.com/daniel-frak/keycloak-user-migration/releases/tag/4.0.0)                                                                                                                                           |
+| 25.X             | [5.0.0](https://github.com/daniel-frak/keycloak-user-migration/releases/tag/5.0.0)                                                                 |
```
Comment (daniel-frak) [inline README.md:25]: This version should just be be `SNAPSHOT` (with no links). The release workflow will change it to the correct tag automatically and that way we won't have dead links.

**Review by daniel-frak — CHANGES_REQUESTED** [review]

**Comment by daniel-frak** [comment]
Thanks!

I just noticed that the end-to-end tests are failing, however :/
It seems that the "should migrate user" test cannot find the `.pf-c-alert__title` element.

Do you fell like you can tackle that as well, or would you like me to help?

**Comment by xgp** [comment]
Looks like a theme change. But sorry to say I have no frontend competency at all. Please help if you have time. Otherwise I ca ask one of my colleagues later this week.

**Comment by daniel-frak** [comment]
Ok, I'll try to carve out some time near the end of this week to take a look.

**Comment by daniel-frak** [comment]
Ok, this is a bit weird. I cloned your repo locally and ran the E2E tests, and they pass.

Looking at the code, the issue is at this place (`docker/e2e/cypress/e2e/migrating_users.cy.js:69`):
```typescript
        cy.get('#kc-forgot-pw-switch').then(($checkbox) => {
            if (!$checkbox.prop('checked')) {
                cy.wrap($checkbox).check({ force: true });
                cy.get('.pf-c-alert__title').should('contain', "Forgot password changed successfully");
            }
        });
```

For whatever reason, when checking the "Forgot password" checkbox in Keycloak's "Realm settings", the confirmation popup is not being detected by Cypress, but seemingly only on CI.

Unfortunately I'm also not too adept in frontend technologies, so we'll have to play this one by ear, a bit.

1) Could you try modifying the test so that it intercepts a GET request, like below? Perhaps that will coax it to behave:

```typescript
        cy.get('#kc-forgot-pw-switch').then(($checkbox) => {
            if (!$checkbox.prop('checked')) {
                cy.intercept('GET', '/admin/realms/master') // Add this
                    .as("reload");
                cy.wrap($checkbox).check({ force: true });
                cy.wait("@reload"); // And this
                cy.get('.pf-c-alert__title').should('contain', "Forgot password changed successfully");
            }
        });
```

2) Could you run the tests locally, as described in [CONTRIBUTING.md](https://github.com/daniel-frak/keycloak-user-migration/blob/master/CONTRIBUTING.md#quick-guide-to-upgrading-keycloak-compatibility)? That way we can make sure that this is strictly a CI issue (and therefore most likely a race condition or something similar).

**Comment by xgp** [comment]
@daniel-frak regarding running the tests locally, I get the same error:
```
DevTools listening on ws://127.0.0.1:41043/devtools/browser/4c62d347-9a73-4a35-9aa3-995824169efa

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.11.0                                                                        │
  │ Browser:        Electron 118 (headless)                                                        │
  │ Node Version:   v20.11.0 (/home/garth/.nvm/versions/node/v20.11.0/bin/node)                    │
  │ Specs:          1 found (migrating_users.cy.js)                                                │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  migrating_users.cy.js                                                           (1 of 1)


  user migration plugin
    1) "before all" hook for "should migrate user"


  0 passing (24s)
  1 failing

  1) user migration plugin
       "before all" hook for "should migrate user":
     AssertionError: Timed out retrying after 20000ms: Expected to find element: `.pf-c-alert__title`, but never found it.

Because this error occurred during a `before all` hook we are skipping the remaining tests in the current suite: `user migration plugin`

Although you have test retries enabled, we do not retry tests when `before all` or `after all` hooks fail
      at Context.eval (webpack://e2e-tests/./cypress/e2e/migrating_users.cy.js:72:45)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        4                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      3                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     23 seconds                                                                       │
  │ Spec Ran:     migrating_users.cy.js                                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /home/garth/projects/saascrud/keycloak-user-migration/docker/e2e/cypress/screens     (1280x720)
     hots/migrating_users.cy.js/user migration plugin -- should migrate user -- befor               
     e all hook (failed).png                                                                        


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  migrating_users.cy.js                    00:23        4        -        1        -        3 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 1 failed (100%)                     00:23        4        -        1        -        3  
```

Inspecting the HTML, it looks like the style has changed (Keycloak team upgraded to Patternfly 5) from `pf-c-alert__title` `pf-v5-c-alert__title`. When I replace all occurrences of that, it passes that error, but then fails with another error:

```
DevTools listening on ws://127.0.0.1:32863/devtools/browser/bb9e67a5-2fca-4309-a779-927dc880b362

====================================================================================================

  (Run Starting)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Cypress:        13.11.0                                                                        │
  │ Browser:        Electron 118 (headless)                                                        │
  │ Node Version:   v20.11.0 (/home/garth/.nvm/versions/node/v20.11.0/bin/node)                    │
  │ Specs:          1 found (migrating_users.cy.js)                                                │
  │ Searched:       cypress/e2e/**/*.cy.{js,jsx,ts,tsx}                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


────────────────────────────────────────────────────────────────────────────────────────────────────
                                                                                                    
  Running:  migrating_users.cy.js                                                           (1 of 1)


  user migration plugin
    1) "before all" hook for "should migrate user"


  0 passing (7s)
  1 failing

  1) user migration plugin
       "before all" hook for "should migrate user":
     CypressError: Timed out retrying after 5000ms: `cy.wait()` timed out waiting `5000ms` for the 1st request to the route: `realm`. No request ever occurred.

https://on.cypress.io/wait

Because this error occurred during a `before all` hook we are skipping the remaining tests in the current suite: `user migration plugin`

Although you have test retries enabled, we do not retry tests when `before all` or `after all` hooks fail
      at cypressErr (http://localhost:8024/__cypress/runner/cypress_runner.js:75613:18)
      at Object.errByPath (http://localhost:8024/__cypress/runner/cypress_runner.js:75668:10)
      at checkForXhr (http://localhost:8024/__cypress/runner/cypress_runner.js:135230:84)
      at <unknown> (http://localhost:8024/__cypress/runner/cypress_runner.js:135256:28)
      at tryCatcher (http://localhost:8024/__cypress/runner/cypress_runner.js:1807:23)
      at Promise.attempt.Promise.try (http://localhost:8024/__cypress/runner/cypress_runner.js:4315:29)
      at whenStable (http://localhost:8024/__cypress/runner/cypress_runner.js:143678:68)
      at <unknown> (http://localhost:8024/__cypress/runner/cypress_runner.js:143619:14)
      at tryCatcher (http://localhost:8024/__cypress/runner/cypress_runner.js:1807:23)
      at Promise._settlePromiseFromHandler (http://localhost:8024/__cypress/runner/cypress_runner.js:1519:31)
      at Promise._settlePromise (http://localhost:8024/__cypress/runner/cypress_runner.js:1576:18)
      at Promise._settlePromise0 (http://localhost:8024/__cypress/runner/cypress_runner.js:1621:10)
      at Promise._settlePromises (http://localhost:8024/__cypress/runner/cypress_runner.js:1701:18)
      at Promise._fulfill (http://localhost:8024/__cypress/runner/cypress_runner.js:1645:18)
      at <unknown> (http://localhost:8024/__cypress/runner/cypress_runner.js:5450:46)
  From Your Spec Code:
      at visitMigrationConfigPage (webpack://e2e-tests/./cypress/e2e/migrating_users.cy.js:96:11)
      at configureMigrationPlugin (webpack://e2e-tests/./cypress/e2e/migrating_users.cy.js:78:8)
      at Context.eval (webpack://e2e-tests/./cypress/e2e/migrating_users.cy.js:42:8)




  (Results)

  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ Tests:        4                                                                                │
  │ Passing:      0                                                                                │
  │ Failing:      1                                                                                │
  │ Pending:      0                                                                                │
  │ Skipped:      3                                                                                │
  │ Screenshots:  1                                                                                │
  │ Video:        false                                                                            │
  │ Duration:     7 seconds                                                                        │
  │ Spec Ran:     migrating_users.cy.js                                                            │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘


  (Screenshots)

  -  /home/garth/projects/saascrud/keycloak-user-migration/docker/e2e/cypress/screens     (1280x720)
     hots/migrating_users.cy.js/user migration plugin -- should migrate user -- befor               
     e all hook (failed).png                                                                        


====================================================================================================

  (Run Finished)


       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  migrating_users.cy.js                    00:07        4        -        1        -        3 │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 1 failed (100%)                     00:07        4        -        1        -        3  

```

**Comment by daniel-frak** [comment]
Weirdly, I'm getting the same error (which I probably should have mentioned) when running your code, unless I increase the wait time of `cy.get('.pf-c-alert__title')` or change the code like I did above.

What happens if you change the code like I did (adding the `@saveRealmSettings` check)? Does it also fail in the same way?

**Comment by xgp** [comment]
@pnzrr Can you look at this?

**Comment by daniel-frak** [comment]
Upon a closer look, it seems that the E2E tests have become generally flaky (even on the `master` branch).

I've improved them somewhat on the main branch. Could you pull from `master` and see if the CI passes this time?

In the meantime, I'll do some more investigating regarding the tests...

**Comment by daniel-frak** [comment]
Ah, shucks :/

I'll try to find some time to confirm if the tests are just flaky here. If that's the case, I'll merge the PR and try to fix the test flakiness later...

**Comment by daniel-frak** [comment]
@xgp Ok, so I took a look at your branch and it seems that the change you made, where you replaced `pf-c-alert__title` with `.pf-v5-c-alert__title` is causing the tests to fail. Reverting that change makes the tests pass, at least locally.

**Comment by xgp** [comment]
I still can't get tests to pass locally, even when reverting that change. I'm somewhat out of my element on these tests. As much as I've tried, I'm really unable to grok what's going on.

**Comment by daniel-frak** [comment]
Yeah, frontend's a bit of a mystery for me as well XD

But it does pass on my machine after the revert, so let's get it to that state and I'll merge it in, to not waste more of your time on the E2E test side.

Then I'll try to find some free time to stabilize the tests on the main branch.

**Pull request merged**